broker/query_plan.go

```diff
@@ -35,75 +46,138 @@ func (bpn *blockingPlanNodeImpl) Add(nodes ...common.BlockingPlanNode) {
-		rhs, ok := mn.children[0].(*MergeNode)
+		rhs, ok := mn.children[1].(common.MergeNode)
 		if !ok {
 			err = utils.StackError(nil, "RHS of avg node must be count node")
 			return
 		}
-		if common.Count != rhs.AggType {
+		if common.Count != rhs.AggType() {
 			err = utils.StackError(nil, "RHS of avg node must be count node")
 			return
 		}
 	}
 
 	childrenResult := make([]queryCom.AQLQueryResult, nChildren)
-	errs := make([]error, nChildren)
+	errs := 0
 	wg := &sync.WaitGroup{}
 	for i, c := range mn.children {
 		wg.Add(1)
-		bpn, ok := c.(common.BlockingPlanNode)
-		if !ok {
-			err = utils.StackError(nil, "expecting BlockingPlanNode")
-			return
-		}
 		go func(i int, n common.BlockingPlanNode) {
 			defer wg.Done()
 			var res queryCom.AQLQueryResult
-			res, err = n.Run()
+			res, err = n.Run(ctx)
 			if err != nil {
 				// err means downstream retry failed
-				errs[i] = utils.StackError(err, "child node failed")
+				utils.GetLogger().With(
+					"error", err,
+				).Error("child node failed")
+				errs++
 				return
 			}
 			childrenResult[i] = res
-		}(i, bpn)
+		}(i, c)
 	}
 	wg.Wait()
 
+	if errs > 0 {
+		err = utils.StackError(nil, fmt.Sprintf("%d errors happend running merge node", errs))
+		return
+	}
+
 	result = childrenResult[0]
 	for i := 1; i < nChildren; i++ {
-		ctx := newResultMergeContext(mn.AggType)
-		result = ctx.run(result, childrenResult[i])
-		if ctx.err != nil {
-			err = ctx.err
+		mergeCtx := newResultMergeContext(mn.aggType)
+		result = mergeCtx.run(result, childrenResult[i])
+		if mergeCtx.err != nil {
+			err = mergeCtx.err
 			return
 		}
 	}
+	return
+}
+
+// ScanNode is a BlockingPlanNode that handles rpc calls to fetch data from datanode
+type ScanNode struct {
+	blockingPlanNodeImpl
+
+	query          query.AQLQuery
+	shardID        uint32
+	topo           topology.Topology
+	dataNodeClient dataCli.DataNodeQueryClient
+}
+
+func (sn *ScanNode) Run(ctx context.Context) (result queryCom.AQLQueryResult, err error) {
+	trial := 0
+	for trial < rpcRetries {
+		trial++
 
+		hosts, routeErr := sn.topo.Get().RouteShard(sn.shardID)
+		if routeErr != nil {
+			utils.GetLogger().With(
+				"error", routeErr,
+				"shard", sn.shardID,
+				"trial", trial).Error("failed to route shard")
+			err = utils.StackError(routeErr, "route shard failed")
+			continue
+		}
+		// pick random routable host
+		rand.Seed(time.Now().UTC().UnixNano())
+		idx := rand.Intn(len(hosts))
+		host := hosts[idx]
+
+		var fetchErr error
+		result, fetchErr = sn.dataNodeClient.Query(ctx, host, sn.query)
+		if fetchErr != nil {
+			utils.GetLogger().With(
+				"error", fetchErr,
+				"shard", sn.shardID,
+				"host", host,
+				"query", sn.query,
+				"trial", trial).Error("failed to fetch from datanode")
+			err = utils.StackError(fetchErr, "fetch from datanode failed")
+			continue
+		}
+		utils.GetLogger().With(
+			"trial", trial,
+			"shard", sn.shardID,
+			"host", host).Info("fetch from datanode succeeded")
+	}
+	if result != nil {
```

Comment:
under what circumstance will a result be non nil and err be nil?

Reply:
if a successful downstream call followed a failure, we clear error here and return the result.